### PR TITLE
Update validation

### LINKED ISSUE #11: Update validation scripts
living [here](https://github.com/epiforecasts/covid19-forecast-hub-europe/tree/main/code/validation) - also taking note of https://github.com/reichlab/covid19-forecast-hub/issues/2594

### PR DESCRIPTION
**Pull request by kathsherratt**
Branch for validation / other minor code changes. Holding as a draft PR for visibility only - continuing to update over next week

Should close #11 when complete

### CODE REVIEW

File: code/validation/test-formatting.py [hunk 1]
```diff
@@ -48,7 +48,12 @@ def filename_match_forecast_date(filename):
             return "ERROR %s forecast filename date %s does match forecast_date column %s" % (
                 filename, file_forecast_date, forecast_date_column)
         else:
-            return None
+            today = datetime.date.today()
+            if (forecast_date_column != (str(today))):
```
Comment (sbfnk) [inline code/validation/test-formatting.py:52]: I think in the US they were going to allow either `today` or `today + 1` (not sure why - I've asked)

File: template/create_eu_codes.R [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Get EU + UK country names + codes
+
+eu28 <- countrycode::codelist
+
+eu28 <- eu28[!is.na(eu28$eu28),]
```
Comment (sbfnk) [inline template/create_eu_codes.R:5]: The plan is to have at least EU + EFTA + UK - see here for getting a list of iso3c codes for the relevant countries
https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/35f5fe94d137a9848ed7a53880ab7ba8e002e9f8/code/auto_download/ecdc_download.r#L12

**Review by sbfnk — COMMENTED** [review]
Great stuff!

File: code/validation/test-formatting.py (hunk 1 above)
Comment (kathsherratt) [reply code/validation/test-formatting.py]: Updated after decision to permit only the Monday submission date in either forecast_date or filename.

**Comment by kathsherratt** [comment]
Validation is ready for review. 
- See [updated-python-files.py](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/5eab47df353da155111c06273d1dd934c5aa03f4/code/validation/adapted-python-files.py) for details with each file's modifications
- Flagging this comment: 
https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/5eab47df353da155111c06273d1dd934c5aa03f4/code/validation/adapted-python-files.py#L40 which would be best checked by someone familiar with Zoltar - can additional columns be stored in Zoltar and used as a filter in queries?

**Comment by kathsherratt** [comment]
Ready for review (@sbfnk) after decision to use JHU data and epiweek. Overview on [code/validation/adapted-python-files.py](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/5eab47df353da155111c06273d1dd934c5aa03f4/code/validation/adapted-python-files.py).

File: code/validation/covid19.py [hunk 3]
```diff
@@ -1,76 +1,55 @@
...
 
+# set valid quantiles
+VALID_QUANTILES = [0.010, 0.025, 0.050, 0.100, 0.150, 0.200, 0.250, 0.300,
+                   0.350, 0.400, 0.450, 0.500, 0.550, 0.600, 0.650, 0.700,
+                   0.750, 0.800, 0.850, 0.900, 0.950, 0.975, 0.990]
 
-VALID_QUANTILES = [0.010, 0.025, 0.050, 0.100, 0.150, 0.200, 0.250, 0.300, 0.350, 0.400, 0.450, 0.500, 0.550, 0.600,
-                   0.650, 0.700, 0.750, 0.800, 0.850, 0.900, 0.950, 0.975, 0.990]  # incoming must be a subset of these
+# set range of valid scenario IDs
+VALID_SCENARIO_ID = ["forecast", "scenario1", "scenario2"] # add new scenarios here
```
Comment (sbfnk) [inline code/validation/covid19.py:24]: looks good - for now let's just define `"forecast"`

File: code/validation/covid19.py [hunk 4]
```diff
@@ -1,76 +1,55 @@
...
 
-FIPS_CODES =  {"Germany": ["GM", "GM01", "GM02", "GM03", "GM04", "GM05", "GM06",
-                            "GM07", "GM08", "GM09", "GM10", "GM11", "GM12",
-                            "GM13", "GM14", "GM15", "GM16"],
-               "Poland": ["PL", "PL72", "PL73", "PL74", "PL75", "PL76", "PL77",
-                           "PL78", "PL79","PL80", "PL81", "PL82", "PL83", 
-                           "PL84", "PL85", "PL86", "PL87"]}
-
+# get location codes as list
+codes = list(pd.read_csv('././template/locations_eu.csv')['iso3c'])
```
Comment (sbfnk) [inline code/validation/covid19.py:12]: Not a big fan of hardcoded relative path names - could consider using pyprojroot or a similar solution.

File: code/validation/covid19.py [hunk 5]
```diff
@@ -81,86 +60,96 @@ def validate_quantile_csv_file(csv_fp, mode, country):
...
+    # 5. Validate date alignment (Sunday-Saturday epi week)
+    weekday_to_sun_based = {i: i + 2 if i != 6 else 1 for i in range(7)}  # Sun=1, Mon=2, ..., Sat=7
+
+    # 5.1 for x week ahead targets, weekday(target_end_date) should be a Sat
+    if weekday_to_sun_based[target_end_date.weekday()] != 7:
+       error_messages.append(f"target_end_date was not a Saturday: {target_end_date}. row={row}")
+       return error_messages  # terminate - depends on valid target_end_date
+   
+    # 5.2 Forecast date should always be Mon
+    if abs(weekday_to_sun_based[forecast_date.weekday()]) != 2:
```
Comment (sbfnk) [inline code/validation/covid19.py:134]: doesn't need the `abs`, right?

File: code/validation/covid19.py [hunk 6]
```diff
@@ -81,86 +60,96 @@ def validate_quantile_csv_file(csv_fp, mode, country):
...
+            error_messages.append(f"Error > forecast_date was not a Monday, row={row}")
+
+    # 5.3 For x week ahead targets, ensure 1-week ahead forecast is for next Sat
+    #   - set exp_target_end_date and then validate it
+    weekday_diff = datetime.timedelta(days=(abs(weekday_to_sun_based[target_end_date.weekday()] -
```
Comment (sbfnk) [inline code/validation/covid19.py:139]: Doesn't need the `abs`, right? We already checked that `weekday_to_sun_based[target_end_date.weekday()]` is 7 so this should be >= 0

File: code/validation/covid19.py [hunk 7]
```diff
@@ -81,86 +60,96 @@ def validate_quantile_csv_file(csv_fp, mode, country):
...
+                                                weekday_to_sun_based[forecast_date.weekday()])))
+    delta_days = weekday_diff + datetime.timedelta(days=(7 * step_ahead_increment))
+    exp_target_end_date = forecast_date + delta_days
+
+    if target_end_date != exp_target_end_date:
+        error_messages.append(f"Error > target_end_date was not the expected Sunday. forecast_date = {forecast_date}, "
```
Comment (sbfnk) [inline code/validation/covid19.py:145]: Saturday?

File: template/generate_template.R [hunk 8]
```diff
@@ -1,60 +1,109 @@
...
+# Weekly forecasts are Sun-Sat with submission (forecast_date) on Mon
 
-# Note that weekly incidence forecasts are always from Sunday (included) to Saturday (included; this is target_end_date)
 
 # define the date on which forecasts are generated:
-forecast_date <- as.Date("2020-07-13")
+forecast_date <- as.Date("2021-02-01")
 
-locations <- read.csv("state_codes_germany.csv")
+locations <- read.csv("template/locations_eu.csv")
```
Comment (sbfnk) [inline template/generate_template.R:14]: can we use `here::here` so it doesn't only work when run from root?

File: template/generate_template.R [hunk 9]
```diff
@@ -1,60 +1,109 @@
...
+  forecast_1_wk_ahead_end = forecast_1_wk_ahead_start + 6)
+write.csv(forecast_date, file = paste0("template/forecast-dates.csv"), row.names = FALSE)
+
+
+# Daily to epiweek conversion ---------------------------------------------
+
+library(magrittr)
+
+epiweeks <- tibble::tibble(
+    date = seq.Date(as.Date("2019-12-29"), as.Date("2022-01-08"), by = 1),
```
Comment (sbfnk) [inline template/generate_template.R:74]: Same again - let's add a year (ever the optimist...)

File: template/metadata-JHU_IDD-CovidSP.txt [hunk 10]
```diff
@@ -0,0 +1,54 @@
+team_name: Johns Hopkins ID Dynamics COVID-19 Working Group
```
Comment (sbfnk) [inline template/metadata-JHU_IDD-CovidSP.txt:1]: I thought we were going with the json format as in the US? Happy to discuss if you think this is more straightforward.

File: template/generate_template.R [hunk 11]
```diff
@@ -1,60 +1,109 @@
...
+dat$forecast_date <- as.Date(dat$forecast_date, origin = "1970-01-01")
+
+write.csv(dat, file = paste0("data-processed/Template-ExampleModel/", forecast_date, "-Template-ExampleModel.csv"), row.names = FALSE)
+write.csv(dat, file = paste0("template/", forecast_date, "-Template-ExampleModel.csv"), row.names = FALSE)
+
+
+# Submission dates & epiweeks ------------------------------------
+
+forecast_date <- tibble::tibble(
+  forecast_date	= seq.Date(as.Date("2021-02-01"), as.Date("2022-01-02"), by = 7),
```
Comment (sbfnk) [inline template/generate_template.R:62]: I'm really uncomfortable with having a fixed end date here as this is bound to be forgotten when it comes to it but I realise this is hardcoded in a few places and will be a pain to change. At the least can we add a year?

File: code/validation/adapted-python-files.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+ - [x] test-formatting-local.py - MODIFIED
+         - removed "COUNTRIES" list, "country" and "mode", not used in filename
+
+ - [x] validate_filenames.py- MODIFIED
+         - removed from filename check: countries, icu, case
+
+ - [x] get_commit_dates.py - UNCHANGED
+
+ - [x] cdc_io.py - UNCHANGED
+     - note: some functions are modified/used in quantile_io.py / covid19.py
```
Comment (sbfnk) [inline code/validation/adapted-python-files.py:55]: can you put here which ones these functions are?

File: code/validation/adapted-python-files.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+
+"""
+23 Jan 2021 - Kath
+Note of files modified for Europe Hub from Germany/Poland Forecast Hub
```
Comment (sbfnk) [inline code/validation/adapted-python-files.py:5]: It would be great to also adapt the file lists in `decisions.md` according to the things listed here

**Review by sbfnk — CHANGES_REQUESTED** [review]
Amazing stuff!

General comments:
- at the moment this just checks that metadata exists; I think we should use the more comprehensive metadata format they're using in the US, with corresponding [validation code](https://github.com/reichlab/covid19-forecast-hub/blob/master/code/validation/validation_functions/metadata.py) - though I'd suggest to merge this PR first and open a separate issue to keep things tidy
- we'll probably use ISO-2 codes, but this should be largely minor changes in `covid19.py` and `locations_eu.csv`
- is the `pip` file needed?

**Comment by sbfnk** [comment]
Ready to merge? We can do metadata and update decision.md (using the very useful overview in code/validation/adapted-python-files.py) separately,

File: template/metadata-JHU_IDD-CovidSP.txt (hunk 10 above)
Comment (sbfnk) [reply template/metadata-JHU_IDD-CovidSP.txt:1]: I had it wrong - format is YAML so this makes sense.

File: code/validation/adapted-python-files.py (hunk 13 above)
Comment (sbfnk) [reply code/validation/adapted-python-files.py:5]: done

**Comment by sbfnk** [comment]
I added a bunch of minor changes based on reviewing this following the discussion in https://github.com/epiforecasts/covid19-forecast-hub-europe/issues/26.

I think this is now ready to be merged and tested.

**Review by sbfnk — APPROVED** [review]

**Pull request merged**